### DISCUSSION TITLE
Hooks response not working unless hard refresh

### DISCUSSION THREAD

**Discussion by shahensaifullah** [sveltejs/kit#7055] [Q&A]
```
hooks.server.js

import * as cookie from 'cookie';
/** @type {import('@sveltejs/kit').Handle} */
export const handle = async ({ event, resolve }) => {
    const cookies = cookie.parse(event.request.headers.get('cookie') || '');

    const response = await resolve(event);

    if(!('csrf_bongo' in cookies)){
        const data = await fetch('http://127.0.0.1:8000/api/auth/user/csrf')
        response.headers.set('set-cookie', data.headers.get("set-cookie"))
    }
    return response;
};
```
Here is my response.header:
```
HeadersList {
  [Symbol(headers map)]: Map(7) {
    'date' => 'Tue, 27 Sep 2022 19:34:33 GMT',
    'content-type' => 'text/plain; charset=utf-8',
    'content-length' => '2',
    'vary' => 'Cookie, Origin',
    'access-control-allow-origin' => '',
    'access-control-allow-credentials' => 'true',
    'set-cookie' => 'csrf_bongo=00fbdd56-721b-4803-b110-bb6a4581c673; expires=Wed, 28 Sep 2022 07:34:34 GMT; path=/; SameSite=Lax'
  },
  [Symbol(headers map sorted)]: null
}
```
for the first time when code run cookies set to browser. Later when i delete that cookie and refresh the browser cookie is not set to browser. If i hard reload the page, the cookie is set to the browser.
So i log the response.headers when the cookie is not set to the browser. here is the result
```
HeadersList {
  [Symbol(headers map)]: Map(4) {
    'content-type' => 'text/html',
    'etag' => '"spux9l"',
    'x-sveltekit-page' => 'true',
    'set-cookie' => 'csrf_bongo=05fbdd56-721b-4803-b110-bb6a4581c673; expires=Wed, 28 Sep 2022 07:41:33 GMT; path=/; SameSite=Lax'
  },
  [Symbol(headers map sorted)]: null
}
```
But set-cookie is not set to the browser. If i hard reload the page, the cookie is set to the browser.

**Comment by teemingc** [comment]
Which browser are you using? Sometimes it's a visual bug, and switching to a different tab in the web inspector will show the updated list of cookies.

Also, you can now use the SvelteKit Cookies API, which uses the `cookie` package under the hood.
```diff
-   import * as cookie from 'cookie';
/** @type {import('@sveltejs/kit').Handle} */
export const handle = async ({ event, resolve }) => {
-   const cookies = cookie.parse(event.request.headers.get('cookie') || '');
+   const csrf_bongo = event.cookies.get('csrf_bongo');

    const response = await resolve(event);

-   if(!('csrf_bongo' in cookies)){
+   if(!csrf_bongo){
        const data = await fetch('http://127.0.0.1:8000/api/auth/user/csrf')
        response.headers.set('set-cookie', data.headers.get("set-cookie"))
    }
    return response;
};
```
https://kit.svelte.dev/docs/types#sveltejs-kit-cookies

**Reply by shahensaifullah** [reply]
i tried with microsoft edge and chrome browsers with switching tabs but the result is the same with your solution. it return undefined(in console log) unless i hard reload the page. if i hard reload the page, i can see my cookie in my console.
Though cookie is not set to browser, code is working and send request when csrf is not available as well as response.headers set the cookie to its header but somehow not set to the browser unless hard reload.
```
HeadersList {
  [Symbol(headers map)]: Map(4) {
    'content-type' => 'text/html',
    'etag' => '"spux9l"',
    'x-sveltekit-page' => 'true',
    'set-cookie' => 'csrf_bongo=2edbf769-b97b-4f14-ab7c-1a7af6507405; expires=Wed, 28 Sep 2022 18:04:23 GMT; path=/; SameSite=Lax'
  },
  [Symbol(headers map sorted)]: null
}
```
this is the response header after i set my cookie to header when csrf is not available in my condition.

This might be problem of framework caching.

**Reply by teemingc** [reply]
Can you open a new issue and provide a reproduction?
I'm curious as well and would like to find the cause of the problem.

**Reply by teemingc** [reply]
> Later when i delete that cookie and refresh the browser cookie is not set to browser. If i hard reload the page, the cookie is set to the browser.

Can you clarify the difference between refresh and hard reloading the page?

**Reply by shahensaifullah** [reply]
![image](https://user-images.githubusercontent.com/44544213/192787387-836b15f2-805a-43fa-86f1-669019203270.png)
this is a hard reload where the browser removes all cached data and reloads the page like when you enter a website first time. but cookies remain there.

**Reply by shahensaifullah** [reply]
> > Later when i delete that cookie and refresh the browser cookie is not set to browser. If i hard reload the page, the cookie is set to the browser.
> 
> Can you clarify the difference between refresh and hard reloading the page?

Could you please, rephase this problem? Set a cookie in your hooks.server.js and check if cookies is set to your browser and remove that cookie from browser, again reload the browser to check if cookie set to browser again no not.

**Reply by teemingc** [reply]
Thank you for explaining hard refreshes. I've learnt something new about Chrome :)

After further investigation, I've submitted an issue regarding this bug #7066
In the meantime, the only workaround is to call `invalidate` to "reset the cache". This is similar to what you were doing with the hard refreshes.
It makes me think it is a cache issue with the way the sveltekit client handles responses.

**Reply by shahensaifullah** [reply]
great. Sveltekit is the future of the javascript framework. You all doing great <3

**Comment by malalecherocks** [comment]
I would say the problem is that you are using hooks.server.js which only it's executed on server calls? such as a hard refresh or calling from the client with a fetch(...)

But I'm a noobie so take what I said with a grain of salt

**Reply by teemingc** [reply]
~`hooks.server.js` runs on every request, so that includes anytime the page is refreshed normally, or even navigating to different pages.~

I only really learned this by console logging the url in the `handle` function and seeing it log on every navigation. The docs mention it runs on every request but it was difficult to picture at the time.

**Reply by malalecherocks** [reply]
Are you sure??? I have this code...

```
export async function handle({ event, resolve }) {
	console.log('hooks.server.js handle()',event.url.pathname);
```
	
And if I navigate I don't see that in the terminal... only when hard refresh

**Reply by malalecherocks** [reply]
I have done an npm update and I can tell you that I don't see that in the terminal when I navigate in the client...

**Reply by malalecherocks** [reply]
I also did a fresh installation with the demo app, and I only added the console.log() to their current hooks.server.js and I can say that it only runs when hard refresh and when a fetch() is executed from client

**Reply by teemingc** [reply]
Whoops, I stand corrected. Thanks for sharing this!

I made a quick repro to test the behaviour of `hooks.server.js`:
https://stackblitz.com/edit/sveltejs-kit-template-default-wnshrp?file=src/hooks.server.js

I'd summarise it as running during SSR (normal refreshes), form actions, server loads and endpoints, etc.

**Reply by malalecherocks** [reply]
Oh I just saw you corrected it... this is important for me because I'm trying to resolve this similar problem: [(7082)](https://github.com/sveltejs/kit/discussions/7082)

Basically trying to prevent user from navigating if no cookie exists...

**Comment by danawoodman** [comment]
`hooks.server.js` only runs on server requests (initial SSR, API calls, form actions), hence the name. 

You will need client side logic to invalidate a session when a cookie expires/is invalidated. You could have an endpoint that you hit which returns the user or a 401 response or something of the sort, then respond accordingly. 

A few ways you can accomplish this:

- Poll every X minutes
- Poll on navigation (probably by hooking into `afterNavigate` in your layout)
- Check on any client request and redirect them when one fails with a 401 response
- Or better yet a combination of all 3

Also, have a look at https://github.com/pilcrowOnPaper/lucia-sveltekit which implements client side invalidation.

**Reply by teemingc** [reply]
I think the larger issue is that refreshing the page (running hooks.server.js) does not set the cookie in the response if it has been deleted via browser dev tools / client-side javascript. Is this normal?

**Reply by danawoodman** [reply]
I don't fully understand. If you delete a cookie via DevTools it is gone, thus when reloading the server won't have access to it anymore. You can't delete server side cookies from the client AFAIK as they're effectively hidden from the client. Are you setting the cookie on the client and not the server?

**Reply by teemingc** [reply]
Sorry, for the confusion. I didn't know there's a difference between server-side and client-side cookies.

I've filed an issue for this a few days ago in #7066 
My understanding was that: if the browser receives a `Response` that has a `set-cookie` header, the browser sets the cookie.

1. A `set-cookie` header is added to the `Response` in `hooks.server.js`.
2. The cookie is stored in the browser after the initial request.
3. The cookie is deleted by opening `Application > Cookies` in Chrome Dev tools and deleting it or using JavaScript and accessing `document.cookie`.

The issue:
1. The cookie does not get set again after refreshing the page.
3. It is only set again if browser cache is cleared or after `invalidate` is called.

Is it expected behaviour that the cookie does not get set again under these conditions?

**Reply by danawoodman** [reply]
Couple things:


- If you have the httpOnly flag set the browser would set the cookie but JavaScript on your page would not be able to read/change the cookie
- You don't need an external library to parse/set cookies, just use `request.cookies`
- You should use `request.cookies` to set the cookie as this will set httpOnly and secure flags. Right now you're setting a cookie that is visible on the client, which I don't think you want?
- Setting a cookie will set it in the browser (test this by hitting the refresh button) under DevTools > Application > Cookies
- Are you trying to read the cookie on the client?

**Reply by shahensaifullah** [reply]
my server response a csrf cookie token with expire time 10s. my frontend logic says, when csrf cookie token(not httponly) not found in browser, send a request to server and server response a cookie with "set-cookie" header which is automatically set the cookie to browser.
but problem is after 10s after deleting the cookie from client automatically, client send the request to server but svelte response donot set that cookie to client unless use hard reload with clear cache of browser.
and obviously i need to read cookie from my client because i use custom header to set my csrf token.

1. /** @type {import('@sveltejs/kit').Handle} */
2. export const handle = async ({ event, resolve }) => {
3.    const csrf_bongo = event.cookies.get('csrf_bongo');
4. 
5.     const response = await resolve(event);
6.    if(!csrf_bongo){
7.         const data = await fetch('http://127.0.0.1:8000/api/auth/user/csrf')
9.         response.headers.set('set-cookie', data.headers.get("set-cookie"))
10.    }
11.    console.log(response.headers)
12.     return response;
13. };

but after expired the cookie, my logic send request to server and get a cookie and i can see the "set-cookie" in line 10. console, return response not set the cookie to browser unless i hard reload with clear cache to browser.
so it might be caching problem. Svelte might cache cookie in my browser and donot remove that cache when cookie expire..

**Reply by danawoodman** [reply]
Why not just put the CSRF token in a hidden field in the form and not rely on cookies?

Seems pretty fragile, especially with a 10s expiry. 

Instead consider attaching the CSRF to the user session and return that in your Svelte file and load them in your forms:

```html
<input type="hidden" name="csrf" value={user.csrf} />
```

Another option would be to encrypt a token with a time code and decrypt it and [allow tokens created in a given range](https://softwareengineering.stackexchange.com/a/140165)

Also, Im no security expert, but having it in a cookie seems problematic because it gets sent when making requests via the browser, so it could open the token up to get used by the attacker. 

PS your code formatting is completely broken, try wrapping it in triple-backticks

**Reply by shahensaifullah** [reply]
> Why not just put the CSRF token in a hidden field in the form and not rely on cookies?
> 
> Seems pretty fragile, especially with a 10s expiry.
> 
> Instead consider attaching the CSRF to the user session and return that in your Svelte file and load them in your forms:
> 
> ```
> <input type="hidden" name="csrf" value={user.csrf} />
> ```

set 10s only for check what happened after removing the cookie from the browser.
but where is user? there is no user there. This token set for everyone, not only for user. This is not issue to set token user or not. issue is svelte is not set cookie which is from server. csrf always set to custom header.
problem is, svelte initially set the cookie to browser, later, after expired the cookie, why svelte not cookie set to browser?

**Reply by shahensaifullah** [reply]
my analysis says svelte caching is prevent to set of cookies again.

**Reply by danawoodman** [reply]
Can you create a basic reproduction reproduction for me to look at?

**Reply by shahensaifullah** [reply]
> Can you create a basic reproduction reproduction for me to look at?

https://github.com/sveltejs/kit/issues/7066#issue-1389511726
here is a repo where you can look over it.
